File: python/Intervals/min_interval_include_each_query.py
from typing import List
import heapq

class Solution:
    def minInterval(self, intervals: List[List[int]], queries: List[int]) -> List[int]:
        """
        For each query, find the length of the smallest interval from intervals that contains the query.
        If no such interval exists, return -1 for that query.
        
        Approach:
        - Sort intervals by their start point.
        - For each query (in sorted order), add all intervals whose start <= query to a min-heap (by interval length).
        - Remove from the heap any intervals whose end < query (they can't cover the query).
        - The top of the heap (if any) is the smallest interval covering the query.
        - Store results in a dictionary for each query, then return results in the original query order.
        
        Time Complexity: O((N + Q) log N), where N = len(intervals), Q = len(queries)
        - Sorting intervals: O(N log N)
        - Sorting queries: O(Q log Q)
        - Each interval is pushed/popped from the heap at most once: O(N log N)
        - Each query processes the heap: O(Q log N)
        
        Space Complexity: O(N + Q)
        - Heap can store up to N intervals
        - Result dictionary and output list store up to Q results
        
        Args:
            intervals: List[List[int]] - List of intervals [start, end]
            queries: List[int] - List of query points
        Returns:
            List[int] - For each query, the length of the smallest covering interval or -1
        """
        intervals.sort()
        
        minHeap = []  # Min-heap of (interval length, interval end)
        res, i = {}, 0

        for q in sorted(queries):
            # Add all intervals starting before or at q
            while i < len(intervals) and intervals[i][0] <= q:
                l, r = intervals[i]
                heapq.heappush(minHeap, (r - l + 1, r))
                i += 1
            
            # Remove intervals that end before q
            while minHeap and minHeap[0][1] < q:
                heapq.heappop(minHeap)
            # The top of the heap is the smallest interval covering q
            res[q] = minHeap[0][0] if minHeap else -1
        
        # Return results in the original query order
        return [res[q] for q in queries]

# -------------------
# Test Cases
# -------------------
def test_min_interval():
    s = Solution()
    # Test case 1: Basic case
    intervals = [[1,4],[2,4],[3,6],[4,4]]
    queries = [2,3,4,5]
    assert s.minInterval(intervals, queries) == [3, 3, 1, 4], "Failed on basic case"
    # Test case 2: No interval covers query
    intervals = [[1,2],[3,4]]
    queries = [0,5]
    assert s.minInterval(intervals, queries) == [-1, -1], "Failed on no coverage"
    # Test case 3: All intervals cover all queries
    intervals = [[1,10],[2,9],[3,8]]
    queries = [4,5,6]
    assert s.minInterval(intervals, queries) == [6, 6, 6], "Failed on all cover"
    # Test case 4: Overlapping intervals, different lengths
    intervals = [[1,5],[2,6],[3,7]]
    queries = [4]
    assert s.minInterval(intervals, queries) == [5], "Failed on overlapping intervals"
    # Test case 5: Single interval, single query
    intervals = [[1,1]]
    queries = [1]
    assert s.minInterval(intervals, queries) == [1], "Failed on single interval/query"
    # Test case 6: Query at interval boundary
    intervals = [[1,3],[2,4]]
    queries = [1,4]
    assert s.minInterval(intervals, queries) == [3, 3], "Failed on boundary queries"
    # Test case 7: Empty intervals
    intervals = []
    queries = [1,2,3]
    assert s.minInterval(intervals, queries) == [-1, -1, -1], "Failed on empty intervals"
    print("All test cases passed!")

if __name__ == "__main__":
    test_min_interval()

File: C++/Intervals/min_interval_include_each_query.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> minInterval(vector<vector<int>>& intervals, vector<int>& queries) {
        /*
         * Approach:
         * - Sort the intervals by their start point.
         * - Create a vector of (query, original index) pairs and sort by query value.
         * - Use a min-heap (priority_queue) to keep track of intervals that can cover the current query.
         *   The heap stores (interval length, interval end).
         * - For each query:
         *      - Add all intervals that start <= query to the heap.
         *      - Remove from the heap all intervals that end < query.
         *      - If the heap is not empty, the smallest interval covering the query is at the top.
         *      - If no interval covers the query, answer is -1.
         * - Return answers in the original query order.
         *
         * Time Complexity: O((N+Q) log N), where N = number of intervals, Q = number of queries.
         * Space Complexity: O(N+Q), for heap and result storage.
         */

        // Sort intervals by their start point
        sort(intervals.begin(), intervals.end());

        // Pair each query with its index to restore original order later
        vector<pair<int, int>> sortedQueries;
        for (int i = 0; i < queries.size(); ++i) {
            sortedQueries.push_back({queries[i], i});
        }
        sort(sortedQueries.begin(), sortedQueries.end());

        // Result vector initialized with -1
        vector<int> result(queries.size(), -1);

        // Min-heap: (interval length, interval end)
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> minHeap;

        int i = 0; // pointer for intervals

        // Loop over queries in sorted order
        for (auto& p : sortedQueries) {
            int q = p.first;   // query value
            int idx = p.second; // original index of the query

            // Add all intervals starting before or at q
            while (i < intervals.size() && intervals[i][0] <= q) {
                int l = intervals[i][0];
                int r = intervals[i][1];
                minHeap.push({r - l + 1, r});
                ++i;
            }

            // Remove intervals that end before q
            while (!minHeap.empty() && minHeap.top().second < q) {
                minHeap.pop();
            }

            // The smallest interval covering q is at the top of the heap
            if (!minHeap.empty()) {
                result[idx] = minHeap.top().first;
            }
        }

        return result;
    }
};

/*
 * Line-by-line syntax explanation:
 *
 * #include <bits/stdc++.h>               → Includes all standard C++ libraries.
 * using namespace std;                  → So we don’t need to write std:: before everything.
 *
 * class Solution { … };                 → Defines the Solution class with public methods.
 *
 * vector<int> minInterval( … )         → Function that takes intervals & queries and returns answers.
 *
 * sort(intervals.begin(), intervals.end());
 *                                        → Sorts intervals by start time.
 *
 * vector<pair<int,int>> sortedQueries; → Vector of pairs to store queries with their original indices.
 *
 * for (…) sortedQueries.push_back(…); → Build the query-index pairs.
 *
 * sort(sortedQueries.begin(), sortedQueries.end());
 *                                        → Sort queries by query value.
 *
 * priority_queue<pair<int,int>, vector<pair<int,int>>, greater<pair<int,int>>> minHeap;
 *                                        → Min-heap of (interval length, interval end).
 *
 * int i = 0;                            → Pointer to iterate over intervals.
 *
 * for (auto& p : sortedQueries) {      → Loop through each query.
 *
 * int q = p.first, idx = p.second;    → Extract query value and its original index.
 *
 * while (i < intervals.size() && intervals[i][0] <= q) …
 *                                        → Add intervals starting before or at q into the heap.
 *
 * while (!minHeap.empty() && minHeap.top().second < q) …
 *                                        → Remove intervals that end before q.
 *
 * if (!minHeap.empty()) result[idx] = minHeap.top().first;
 *                                        → If the heap has valid intervals, pick the smallest length.
 *
 * return result;                        → Return the final answers.
 *
 */

